pymodaq_plugins/daq_viewer_plugins/plugins_1D/daq_1Dviewer_LecroyWaverunner6Zi.py

```diff
@@ -1,202 +1,223 @@
+                    raise Exception('no controller has been defined externally while this detector is a slave one')
+                else:
+                    self.controller = controller
+            else:
+                self.controller = active_dso
+                usb_address = "USBTMC:" + self.settings.child(('VISA_ressources')).value()
+                self.controller.MakeConnection(usb_address)
+                # set the timeout of the scope to 10 seconds
+                # may be not needed
+                self.controller.setTimeout(10)
+
+            self.status.initialized = True
+            self.status.controller = self.controller
+            # self.controller.WriteString(r"""vbs 'app.acquisition.triggermode = "stopped" ' """, 1)
+            return self.status
+        except Exception as e:
+            self.emit_status(ThreadCommand('Update_Status', [getLineInfo() + str(e), 'log']))
+            self.status.info = getLineInfo() + str(e)
+            self.status.initialized = False
+            return self.status
+
+    def commit_settings(self, param):
+        """
+            Activate the parameters changes in the hardware.
+
+            =============== ================================= ============================
+            **Parameters**   **Type**                         **Description**
+            *param*         instance of pyqtgraph.parameter   The parameter to be checked.
+            =============== ================================= ============================
+
+            See Also
+            --------
+            daq_utils.ThreadCommand
+        """
+        try:
+            pass
+        except Exception as e:
+            self.emit_status(ThreadCommand('Update_Status',[getLineInfo() + str(e),'log']))
+
+    def grab_data(self, Naverage=1, **kwargs):
+        """
+            | Start new acquisition.
+            | Grab the current values.
+            | Send the data_grabed_signal once done.
+
+            =============== ======== ===============================================
+            **Parameters**  **Type**  **Description**
+            *Naverage*      int       Number of values to average
+            =============== ======== ===============================================
+        """
+        channel = self.settings.child(('channels')).value()['selected']
+
+        self.controller.WriteString(r"""vbs? 'return=app.acquisition.horizontal.numsegments' """, 1)
+        numSegments = self.controller.ReadString(8)
+
+        self.controller.WriteString(r"""vbs? 'return=app.acquisition.horizontal.samplemode' """, 1)
+        sampleMode = self.controller.ReadString(8)
+
+        # The WaitForOPC method is used to wait for previous commands to be interpreted before continuing
+        # It may be not needed here
+        if not self.controller.WaitForOPC():
+            raise Exception("Wait for OPC error")
+
+        waveform = self.controller.GetScaledWaveformWithTimes(channel[0], 1e8, 0)

```

Comment:
I would get the waveform after having waited for acquiredSegments == numSegments no ?